## Cleaning Data

In this notebook we start the analizing the two dataset and correct errors


In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

## Taxi Dataset

We start from the taxi dataset

In [2]:
df = pd.read_csv(Path("data/Chicago_Taxi_Trips_2020_10percent.csv")) # this open the csv dataset considering the path
#print(df.head()) #first 5 lines

df.columns = [c.replace(' ', '_') for c in df.columns]  #replaces spaces in columns with underscores

In [3]:
df.head()

Trip_ID  \
0  7fa9327d2a62e0f1f9eac0f5a4ca0bb5338c22b2   
1  493354c5b05333aeeb05055a4914c5d38bacef60   
2  450f3aa6b31aa37cbb7c972f92a9ba31ab506c79   
3  39100cdb94f6da6b54f71f1d1cdc2ba330f97e09   
4  36bbb8cf2622b3ce7dde8f6bbaacc903a8403333   

                                             Taxi_ID    Trip_Start_Timestamp  \
0  03906d62f91d139ab93f74f50d1b208805b7fbb61e8d1e...  01/01/2020 12:00:00 AM   
1  ca885a78545e9a44a5f365164cfbaca32fee95082b0d4f...  01/01/2020 12:00:00 AM   
2  81bc2020b8536280419b3e152ad07a0c5b9f53cbb503a0...  01/01/2020 12:00:00 AM   
3  a548c63608ec50e02a2311a0b6b1b2274995e3637ccb4c...  01/01/2020 12:00:00 AM   
4  33164e16dd29b1c58cd15cce31df4bfcb75d9903cb66de...  01/01/2020 12:00:00 AM   

       Trip_End_Timestamp  Trip_Seconds  Trip_Miles  Pickup_Census_Tract  \
0  01/01/2020 12:00:00 AM         178.0        0.48         1.703132e+10   
1  01/01/2020 12:30:00 AM        1200.0        5.30                  NaN   
2  01/01/2020 12:15:00 AM         540.0        1.70         1.703108e+10   
3  01/01/2020 12:15:00 AM         419.0        0.98         1.703128e+10   
4  01/01/2020 12:30:00 AM        1800.0        1.50         1.703128e+10   

   Dropoff_Census_Tract  Pickup_Community_Area  Dropoff_Community_Area  ...  \
0          1.703132e+10                   32.0                    32.0  ...   
1                   NaN                    8.0                    23.0  ...   
2          1.703108e+10                    8.0                     8.0  ...   
3          1.703184e+10                   28.0                    32.0  ...   
4          1.703132e+10                   28.0                    32.0  ...   

   Extras  Trip_Total  Payment_Type                    Company  \
0     3.0       10.00   Credit Card                  Flash Cab   
1     0.0       20.10   Credit Card  Star North Management LLC   
2     0.0        8.00          Cash  Taxi Affiliation Services   
3     1.0        7.50          Cash               City Service   
4     1.0       15.25          Cash           Medallion Leasin   

   Pickup_Centroid_Latitude Pickup_Centroid_Longitude  \
0                 41.877406                -87.621972   
1                 41.899602                -87.633308   
2                 41.907492                -87.635760   
3                 41.885300                -87.642808   
4                 41.885300                -87.642808   

               Pickup_Centroid_Location  Dropoff_Centroid_Latitude  \
0  POINT (-87.6219716519 41.8774061234)                  41.884987   
1   POINT (-87.6333080367 41.899602111)                  41.900070   
2  POINT (-87.6357600901 41.9074919303)                  41.892073   
3  POINT (-87.6428084655 41.8853000224)                  41.880994   
4  POINT (-87.6428084655 41.8853000224)                  41.884987   

   Dropoff_Centroid_Longitude            Dropoff_Centroid__Location  
0                  -87.620993  POINT (-87.6209929134 41.8849871918)  
1                  -87.720918  POINT (-87.7209182385 41.9000696026)  
2                  -87.628874  POINT (-87.6288741572 41.8920726347)  
3                  -87.632746  POINT (-87.6327464887 41.8809944707)  
4                  -87.620993  POINT (-87.6209929134 41.8849871918)  

[5 rows x 23 columns]

In [4]:
df.shape

(342292, 23)

### Data preparation - Time

In [5]:
#This code recognises AM/PM difference
df.Trip_Start_Timestamp = pd.to_datetime(df.Trip_Start_Timestamp, format='%m/%d/%Y %I:%M:%S %p')
df.Trip_End_Timestamp = pd.to_datetime(df.Trip_End_Timestamp, format='%m/%d/%Y %I:%M:%S %p')
#This line excludes trips that End before Starting (90 trips)
df = df[df.Trip_End_Timestamp >= df.Trip_Start_Timestamp]

In [6]:
#TODO ci sarebbe da controllare se la durata in secondi sia la stessa della differenza fra partenza e arrivo
#df=df[(df.Trip_End_Timestamp-df.Trip_Start_Timestamp)==pd.to_timedelta(df.Trip_Seconds * 10 ** 9)]

In [7]:
df.shape

(342202, 23)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 342202 entries, 0 to 342291
Data columns (total 23 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Trip_ID                     342202 non-null  object        
 1   Taxi_ID                     342184 non-null  object        
 2   Trip_Start_Timestamp        342202 non-null  datetime64[ns]
 3   Trip_End_Timestamp          342202 non-null  datetime64[ns]
 4   Trip_Seconds                342158 non-null  float64       
 5   Trip_Miles                  342201 non-null  float64       
 6   Pickup_Census_Tract         174058 non-null  float64       
 7   Dropoff_Census_Tract        172967 non-null  float64       
 8   Pickup_Community_Area       318231 non-null  float64       
 9   Dropoff_Community_Area      310704 non-null  float64       
 10  Fare                        342168 non-null  float64       
 11  Tips                        342168 non-

In [9]:
df.head()

Trip_ID  \
0  7fa9327d2a62e0f1f9eac0f5a4ca0bb5338c22b2   
1  493354c5b05333aeeb05055a4914c5d38bacef60   
2  450f3aa6b31aa37cbb7c972f92a9ba31ab506c79   
3  39100cdb94f6da6b54f71f1d1cdc2ba330f97e09   
4  36bbb8cf2622b3ce7dde8f6bbaacc903a8403333   

                                             Taxi_ID Trip_Start_Timestamp  \
0  03906d62f91d139ab93f74f50d1b208805b7fbb61e8d1e...           2020-01-01   
1  ca885a78545e9a44a5f365164cfbaca32fee95082b0d4f...           2020-01-01   
2  81bc2020b8536280419b3e152ad07a0c5b9f53cbb503a0...           2020-01-01   
3  a548c63608ec50e02a2311a0b6b1b2274995e3637ccb4c...           2020-01-01   
4  33164e16dd29b1c58cd15cce31df4bfcb75d9903cb66de...           2020-01-01   

   Trip_End_Timestamp  Trip_Seconds  Trip_Miles  Pickup_Census_Tract  \
0 2020-01-01 00:00:00         178.0        0.48         1.703132e+10   
1 2020-01-01 00:30:00        1200.0        5.30                  NaN   
2 2020-01-01 00:15:00         540.0        1.70         1.703108e+10   
3 2020-01-01 00:15:00         419.0        0.98         1.703128e+10   
4 2020-01-01 00:30:00        1800.0        1.50         1.703128e+10   

   Dropoff_Census_Tract  Pickup_Community_Area  Dropoff_Community_Area  ...  \
0          1.703132e+10                   32.0                    32.0  ...   
1                   NaN                    8.0                    23.0  ...   
2          1.703108e+10                    8.0                     8.0  ...   
3          1.703184e+10                   28.0                    32.0  ...   
4          1.703132e+10                   28.0                    32.0  ...   

   Extras  Trip_Total  Payment_Type                    Company  \
0     3.0       10.00   Credit Card                  Flash Cab   
1     0.0       20.10   Credit Card  Star North Management LLC   
2     0.0        8.00          Cash  Taxi Affiliation Services   
3     1.0        7.50          Cash               City Service   
4     1.0       15.25          Cash           Medallion Leasin   

   Pickup_Centroid_Latitude Pickup_Centroid_Longitude  \
0                 41.877406                -87.621972   
1                 41.899602                -87.633308   
2                 41.907492                -87.635760   
3                 41.885300                -87.642808   
4                 41.885300                -87.642808   

               Pickup_Centroid_Location  Dropoff_Centroid_Latitude  \
0  POINT (-87.6219716519 41.8774061234)                  41.884987   
1   POINT (-87.6333080367 41.899602111)                  41.900070   
2  POINT (-87.6357600901 41.9074919303)                  41.892073   
3  POINT (-87.6428084655 41.8853000224)                  41.880994   
4  POINT (-87.6428084655 41.8853000224)                  41.884987   

   Dropoff_Centroid_Longitude            Dropoff_Centroid__Location  
0                  -87.620993  POINT (-87.6209929134 41.8849871918)  
1                  -87.720918  POINT (-87.7209182385 41.9000696026)  
2                  -87.628874  POINT (-87.6288741572 41.8920726347)  
3                  -87.632746  POINT (-87.6327464887 41.8809944707)  
4                  -87.620993  POINT (-87.6209929134 41.8849871918)  

[5 rows x 23 columns]

### Data preparation - MV



We count the number of missing values for each columns

In [10]:
df.isnull().sum()

Trip_ID                            0
Taxi_ID                           18
Trip_Start_Timestamp               0
Trip_End_Timestamp                 0
Trip_Seconds                      44
Trip_Miles                         1
Pickup_Census_Tract           168144
Dropoff_Census_Tract          169235
Pickup_Community_Area          23971
Dropoff_Community_Area         31498
Fare                              34
Tips                              34
Tolls                             34
Extras                            34
Trip_Total                        34
Payment_Type                       0
Company                            0
Pickup_Centroid_Latitude       23919
Pickup_Centroid_Longitude      23919
Pickup_Centroid_Location       23919
Dropoff_Centroid_Latitude      30855
Dropoff_Centroid_Longitude     30855
Dropoff_Centroid__Location     30855
dtype: int64

Using the description given from the City of Chicago we can assume that there are
from 23930 to 30916 trips from (or to) locations outside Chicago.

We firstly check if there are totally empty rows. We will remove them.

In [11]:
df.dropna(how='all').shape

(342202, 23)

Turns out that the all rows have at least one valid value

### Statistics

In [12]:
from scipy.stats import median_abs_deviation #as median_abs_deviation

The following function will help us to compute some statics useful to analyze our data.
???Or correct NaN value using the Winsorizing approach with a reasonable approach to the correct percentile.

In [13]:
def statistics(variable):
    stats = {}
    stats['mean'] = variable.mean()
    stats['std'] = variable.std()
    stats['max'] = variable.max()
    stats['min'] = variable.min()
    stats['median'] = np.nanmedian(variable)  #NANmedian invece di median
    stats['1%'] = np.nanpercentile(variable, 1)
    stats['5%'] = np.nanpercentile(variable, 5)
    stats['10%'] = np.nanpercentile(variable,10)  #NANpercentile invece di percentile
    stats['17,5%'] = np.nanpercentile(variable,17.5)
    stats['25%'] = np.nanpercentile(variable,25)  #NANpercentile invece di percentile
    stats['75%'] = np.nanpercentile(variable,75)  #NANpercentile invece di percentile
    stats['95%'] = np.nanpercentile(variable, 95)
    stats['99%'] = np.nanpercentile(variable,99)  #NANpercentile invece di percentile
    stats['99.5%'] = np.nanpercentile(variable,99.5)  #NANpercentile invece di percentile
    stats['99.9%'] = np.nanpercentile(variable,99.9)  #NANpercentile invece di percentile
    stats['mad'] = median_abs_deviation(variable, nan_policy = 'omit')
    return stats

#### Starting from Trip Seconds...

In [14]:
stat_Trip_Seconds = statistics(df.Trip_Seconds)
stat_Trip_Seconds

{'mean': 849.9995382250305,
 'std': 1687.5361388391007,
 'max': 85320.0,
 'min': 0.0,
 'median': 560.0,
 '1%': 0.0,
 '5%': 60.0,
 '10%': 180.0,
 '17,5%': 276.0,
 '25%': 346.0,
 '75%': 1020.0,
 '95%': 2280.0,
 '99%': 3600.0,
 '99.5%': 4380.0,
 '99.9%': 13835.686000000103,
 'mad': 280.0}

Max value of the Trip duration (in seconds) is 85320 seconds, or 23.7h. This does not make any sense and we think that is an outlier.

For this columns we consider a Winsorizing or Trimming -
We discard values above 99.5th percentiles since they are over 73 min.

We do not consider values below 10th percentiles because trips below 3 min are meaningless.

In [15]:
#Winsorizing of values below the 10th percentile
df.Trip_Seconds = df.Trip_Seconds.clip(lower = stat_Trip_Seconds['10%'])
#Trimming of values above the 99.5th percentile
df = df[df.Trip_Seconds < stat_Trip_Seconds['99.5%']]

df.shape

(340439, 23)

In [16]:
#Visualizing stats after Winsorizing/Trimming
stat_Trip_Seconds_after = statistics(df.Trip_Seconds)
stat_Trip_Seconds_after

{'mean': 794.5381110859802,
 'std': 671.3930996228495,
 'max': 4379.0,
 'min': 180.0,
 'median': 555.0,
 '1%': 180.0,
 '5%': 180.0,
 '10%': 180.0,
 '17,5%': 274.0,
 '25%': 344.0,
 '75%': 1020.0,
 '95%': 2220.0,
 '99%': 3240.0,
 '99.5%': 3600.0,
 '99.9%': 4140.0,
 'mad': 279.0}

#### Now the Trip Miles

In [17]:
stat_Trip_Miles = statistics(df.Trip_Miles)
stat_Trip_Miles

{'mean': 3.5027847067599516,
 'std': 6.292457658468007,
 'max': 980.5,
 'min': 0.0,
 'median': 1.2,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.1,
 '25%': 0.5,
 '75%': 3.5,
 '95%': 17.07,
 '99%': 20.74,
 '99.5%': 25.5,
 '99.9%': 35.2,
 'mad': 1.0999999999999999}

We see that Max = 980 Miles is clearly an outlier. It's grater than the distance needed to go from Chicago to New York.

Winsorizing: < 25th percentile equal to the 25th percentile because trips are less than 0.5 Miles.

Values above 99.9th percentile are instead discarded (study area is about 40km wide on its longest axis).

In [18]:
#Winsorizing of values below the 25th percentile
df.Trip_Miles = df.Trip_Miles.clip(lower = stat_Trip_Miles['25%'])
#Trimming of values above the 99.9th percentile
df = df[df.Trip_Miles < stat_Trip_Miles['99.9%']]

df.shape

(340096, 23)

In [19]:
#Visualizing stats after Winsorizing/Trimming
stat_Trip_Miles_after = statistics(df.Trip_Miles)
stat_Trip_Miles_after

{'mean': 3.550276392547919,
 'std': 5.14975664389192,
 'max': 35.1,
 'min': 0.5,
 'median': 1.2,
 '1%': 0.5,
 '5%': 0.5,
 '10%': 0.5,
 '17,5%': 0.5,
 '25%': 0.5,
 '75%': 3.5,
 '95%': 17.0,
 '99%': 20.27,
 '99.5%': 24.21,
 '99.9%': 31.288100000000558,
 'mad': 0.7}

#### Fares

In [20]:
Fare_stat = statistics(df.Fare)
Fare_stat

{'mean': 14.84980133268242,
 'std': 77.47432072009028,
 'max': 9004.25,
 'min': 0.0,
 'median': 8.25,
 '1%': 3.25,
 '5%': 3.5,
 '10%': 4.75,
 '17,5%': 5.25,
 '25%': 6.0,
 '75%': 15.75,
 '95%': 43.5,
 '99%': 53.0,
 '99.5%': 63.25,
 '99.9%': 82.25,
 'mad': 3.0}

Winsorizing at 1%: exludes below 1st percentile (which is the Base Fare as written on the Chicago.gov website).
Discard above the 99.9th percentile (higher values are probably caused by outliers from Miles and Trips).

In [21]:
df = df[df.Fare <= 82.25]
df.Fare = df.Fare.clip(lower =  Fare_stat['1%'])

Reanalyze

In [22]:
Fare_stat =statistics(df.Fare)
Fare_stat

{'mean': 13.698266691392483,
 'std': 12.546637624014913,
 'max': 82.25,
 'min': 3.25,
 'median': 8.25,
 '1%': 3.25,
 '5%': 3.5,
 '10%': 4.75,
 '17,5%': 5.25,
 '25%': 6.0,
 '75%': 15.75,
 '95%': 43.5,
 '99%': 52.0,
 '99.5%': 60.75,
 '99.9%': 75.0,
 'mad': 3.0}

#### Tips

In [23]:
statistics(df.Tips)

{'mean': 1.5434855736681425,
 'std': 2.880891326664622,
 'max': 300.0,
 'min': 0.0,
 'median': 0.0,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.0,
 '25%': 0.0,
 '75%': 2.0,
 '95%': 8.44,
 '99%': 11.8,
 '99.5%': 13.4,
 '99.9%': 20.0,
 'mad': 0.0}

 This is a subjetive value of each people so 0$ as tip is possible. We discard onlye above the 99.9th percentile.

In [24]:
df =df[df.Tips <= 20]

reanalyze

In [25]:
statistics(df.Tips)

{'mean': 1.5211593579610732,
 'std': 2.6569586054686827,
 'max': 20.0,
 'min': 0.0,
 'median': 0.0,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.0,
 '25%': 0.0,
 '75%': 2.0,
 '95%': 8.35,
 '99%': 11.62,
 '99.5%': 12.95,
 '99.9%': 16.75,
 'mad': 0.0}

#### Tolls

In [26]:
statistics(df.Tolls)

{'mean': 0.0013600172027300485,
 'std': 0.12062766507245369,
 'max': 50.0,
 'min': 0.0,
 'median': 0.0,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.0,
 '25%': 0.0,
 '75%': 0.0,
 '95%': 0.0,
 '99%': 0.0,
 '99.5%': 0.0,
 '99.9%': 0.0,
 'mad': 0.0}

Watching the percentile and considering that we are in a City and so the Tolls should not appear
we can trimming above the 99.9th percentile

In [27]:
df =df[df.Tolls == 0]

In [28]:
statistics(df.Tolls)

{'mean': 0.0,
 'std': 0.0,
 'max': 0.0,
 'min': 0.0,
 'median': 0.0,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.0,
 '25%': 0.0,
 '75%': 0.0,
 '95%': 0.0,
 '99%': 0.0,
 '99.5%': 0.0,
 '99.9%': 0.0,
 'mad': 0.0}

#### Extras.

In [29]:
statistics(df.Extras)

{'mean': 1.0568394591663517,
 'std': 26.86605453630908,
 'max': 9900.66,
 'min': 0.0,
 'median': 0.0,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.0,
 '25%': 0.0,
 '75%': 1.0,
 '95%': 4.0,
 '99%': 12.0,
 '99.5%': 26.0,
 '99.9%': 42.5,
 'mad': 0.0}

Extras = 0$ is accetable.
We trimming above 99.9th

In [30]:
df = df[df.Extras <= 42.5]

reanalyze

In [31]:
statistics(df.Extras)

{'mean': 0.8754950530097872,
 'std': 2.8179486266648315,
 'max': 42.5,
 'min': 0.0,
 'median': 0.0,
 '1%': 0.0,
 '5%': 0.0,
 '10%': 0.0,
 '17,5%': 0.0,
 '25%': 0.0,
 '75%': 1.0,
 '95%': 4.0,
 '99%': 10.0,
 '99.5%': 24.0,
 '99.9%': 35.5,
 'mad': 0.0}

#### Trip Total

In [32]:
Trip_Total_stat = statistics(df.Trip_Total)

This colums is the sum of the previous one so not mak sense a value below the 1th percentile
so we need to Winsorizing the data below this value

In [33]:
df.Trip_Total = df.Trip_Total.clip(Trip_Total_stat['1%'])

reanalyze

In [34]:
statistics(df.Trip_Total)

{'mean': 16.164382112957423,
 'std': 15.501259178097856,
 'max': 134.25,
 'min': 3.25,
 'median': 9.75,
 '1%': 3.25,
 '5%': 4.0,
 '10%': 5.25,
 '17,5%': 6.25,
 '25%': 7.0,
 '75%': 17.75,
 '95%': 53.8,
 '99%': 66.0,
 '99.5%': 76.6,
 '99.9%': 99.0,
 'mad': 3.75}

In [35]:
statistics(df.Pickup_Community_Area)

{'mean': 27.41607920792079,
 'std': 21.20732017456854,
 'max': 77.0,
 'min': 1.0,
 'median': 28.0,
 '1%': 2.0,
 '5%': 6.0,
 '10%': 8.0,
 '17,5%': 8.0,
 '25%': 8.0,
 '75%': 32.0,
 '95%': 76.0,
 '99%': 77.0,
 '99.5%': 77.0,
 '99.9%': 77.0,
 'mad': 20.0}

In [36]:
statistics(df.Dropoff_Community_Area)

{'mean': 23.20770018006814,
 'std': 18.01902434884715,
 'max': 77.0,
 'min': 1.0,
 'median': 24.0,
 '1%': 2.0,
 '5%': 5.0,
 '10%': 7.0,
 '17,5%': 8.0,
 '25%': 8.0,
 '75%': 32.0,
 '95%': 69.0,
 '99%': 77.0,
 '99.5%': 77.0,
 '99.9%': 77.0,
 'mad': 16.0}

The previous two value are NaN or between 1 and 77 so they make sense

In [37]:
df.shape

(338994, 23)

In [38]:
df.isnull().sum()

Trip_ID                            0
Taxi_ID                           18
Trip_Start_Timestamp               0
Trip_End_Timestamp                 0
Trip_Seconds                       0
Trip_Miles                         0
Pickup_Census_Tract           166005
Dropoff_Census_Tract          166754
Pickup_Community_Area          23369
Dropoff_Community_Area         30222
Fare                               0
Tips                               0
Tolls                              0
Extras                             0
Trip_Total                         0
Payment_Type                       0
Company                            0
Pickup_Centroid_Latitude       23319
Pickup_Centroid_Longitude      23319
Pickup_Centroid_Location       23319
Dropoff_Centroid_Latitude      29587
Dropoff_Centroid_Longitude     29587
Dropoff_Centroid__Location     29587
dtype: int64

We discard all the data error (such as wrong datatime) and eclude outlier data
(with the actual miles seconds trips we can have trip out Chicago with sense)

# Analyzing Weather data
 We now open the dataset of weather to simply control it

In [39]:
weather = pd.read_csv(Path("data/OHare_Airport_Daily_Weather_2020.csv"))

weather.columns = [c.replace(' ', '_') for c in weather.columns]

In [40]:
weather.shape

(295, 40)

In [41]:
weather.head()

STATION                                        NAME  LATITUDE  \
0  USW00094846  CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96019   
1  USW00094846  CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96019   
2  USW00094846  CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96019   
3  USW00094846  CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96019   
4  USW00094846  CHICAGO OHARE INTERNATIONAL AIRPORT, IL US  41.96019   

   LONGITUDE  ELEVATION        DATE  AWND AWND_ATTRIBUTES  PRCP  \
0  -87.93162      201.8  2020-01-01   5.3             ,,W   0.0   
1  -87.93162      201.8  2020-01-02   5.2             ,,W   0.0   
2  -87.93162      201.8  2020-01-03   2.7             ,,W   0.0   
3  -87.93162      201.8  2020-01-04   4.1             ,,W   0.8   
4  -87.93162      201.8  2020-01-05   6.2             ,,W   0.0   

  PRCP_ATTRIBUTES  ...  WT04 WT04_ATTRIBUTES  WT05 WT05_ATTRIBUTES  WT06  \
0        ,,W,2400  ...   NaN             NaN   NaN             NaN   NaN   
1        ,,W,2400  ...   NaN             NaN   NaN             NaN   NaN   
2        ,,W,2400  ...   NaN             NaN   NaN             NaN   NaN   
3        ,,W,2400  ...   NaN             NaN   NaN             NaN   NaN   
4        ,,W,2400  ...   NaN             NaN   NaN             NaN   NaN   

  WT06_ATTRIBUTES  WT08 WT08_ATTRIBUTES  WT09 WT09_ATTRIBUTES  
0             NaN   NaN             NaN   NaN             NaN  
1             NaN   NaN             NaN   NaN             NaN  
2             NaN   NaN             NaN   NaN             NaN  
3             NaN   1.0             ,,W   NaN             NaN  
4             NaN   NaN             NaN   NaN             NaN  

[5 rows x 40 columns]

In [42]:
weather.columns

Index(['STATION', 'NAME', 'LATITUDE', 'LONGITUDE', 'ELEVATION', 'DATE', 'AWND',
       'AWND_ATTRIBUTES', 'PRCP', 'PRCP_ATTRIBUTES', 'SNOW', 'SNOW_ATTRIBUTES',
       'SNWD', 'SNWD_ATTRIBUTES', 'TAVG', 'TAVG_ATTRIBUTES', 'TMAX',
       'TMAX_ATTRIBUTES', 'TMIN', 'TMIN_ATTRIBUTES', 'WDF5', 'WDF5_ATTRIBUTES',
       'WSF5', 'WSF5_ATTRIBUTES', 'WT01', 'WT01_ATTRIBUTES', 'WT02',
       'WT02_ATTRIBUTES', 'WT03', 'WT03_ATTRIBUTES', 'WT04', 'WT04_ATTRIBUTES',
       'WT05', 'WT05_ATTRIBUTES', 'WT06', 'WT06_ATTRIBUTES', 'WT08',
       'WT08_ATTRIBUTES', 'WT09', 'WT09_ATTRIBUTES'],
      dtype='object')

We can se that thet columns are not so legible so we rewrite them
following the documentation of the dataset on Website

In [43]:
weather.columns = [c.replace('AWND', 'Average_Daily_Wind_speed') for c in weather.columns]
weather.columns = [c.replace('PRCP', 'Precipitation') for c in weather.columns]
weather.columns = [c.replace('SNWD', 'Snow_depth') for c in weather.columns]
weather.columns = [c.replace('TMAX', 'Max_temperature') for c in weather.columns]
weather.columns = [c.replace('TMIN', 'Minimum temperature') for c in weather.columns]
weather.columns = [c.replace('SNOW', 'Snowfall') for c in weather.columns]
weather.columns = [c.replace('TAVG', 'Average_temperature') for c in weather.columns]
weather.columns = [c.replace('WDF5', 'Direction_of_fastest_5_second_wind') for c in weather.columns]
weather.columns = [c.replace('WSF5', 'Fastest_5_second_wind_speed') for c in weather.columns]
weather.columns = [c.replace('WT', 'Weather_Type_Value_') for c in weather.columns]

In [44]:
weather.columns

Index(['STATION', 'NAME', 'LATITUDE', 'LONGITUDE', 'ELEVATION', 'DATE',
       'Average_Daily_Wind_speed', 'Average_Daily_Wind_speed_ATTRIBUTES',
       'Precipitation', 'Precipitation_ATTRIBUTES', 'Snowfall',
       'Snowfall_ATTRIBUTES', 'Snow_depth', 'Snow_depth_ATTRIBUTES',
       'Average_temperature', 'Average_temperature_ATTRIBUTES',
       'Max_temperature', 'Max_temperature_ATTRIBUTES', 'Minimum temperature',
       'Minimum temperature_ATTRIBUTES', 'Direction_of_fastest_5_second_wind',
       'Direction_of_fastest_5_second_wind_ATTRIBUTES',
       'Fastest_5_second_wind_speed', 'Fastest_5_second_wind_speed_ATTRIBUTES',
       'Weather_Type_Value_01', 'Weather_Type_Value_01_ATTRIBUTES',
       'Weather_Type_Value_02', 'Weather_Type_Value_02_ATTRIBUTES',
       'Weather_Type_Value_03', 'Weather_Type_Value_03_ATTRIBUTES',
       'Weather_Type_Value_04', 'Weather_Type_Value_04_ATTRIBUTES',
       'Weather_Type_Value_05', 'Weather_Type_Value_05_ATTRIBUTES',
       'Weather_Type_Va

Now we count the Nan value

In [45]:
weather.isnull().sum()

STATION                                            0
NAME                                               0
LATITUDE                                           0
LONGITUDE                                          0
ELEVATION                                          0
DATE                                               0
Average_Daily_Wind_speed                           1
Average_Daily_Wind_speed_ATTRIBUTES                1
Precipitation                                      0
Precipitation_ATTRIBUTES                           0
Snowfall                                           1
Snowfall_ATTRIBUTES                                1
Snow_depth                                         0
Snow_depth_ATTRIBUTES                              0
Average_temperature                                1
Average_temperature_ATTRIBUTES                     1
Max_temperature                                    0
Max_temperature_ATTRIBUTES                         0
Minimum temperature                           

We found that the are a lot of NaN value in the Weather
Type columns.

This follow the documentation of the datest because
the Weather Type Value can not be true for all type of weather.

In [46]:
weather.dropna(how='all').shape

(295, 40)

So we can assume that there are not errors in the dataset of weather

Now We generate the csv file data to better preserve the cleaning operation for
visualization

In [47]:
df.to_csv(Path("generated_data/chicago_taxi_small_and_cleaned.csv"), index = False)

In [48]:
weather.to_csv(Path("generated_data/daily_weather_cleaned_and_reformated.csv"), index = False)
